// uniforms/es5/AutoForm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutoForm = exports.Auto = void 0;
var tslib_1 = require("tslib");
var clone_1 = tslib_1.__importDefault(require("lodash/clone"));
var isEqual_1 = tslib_1.__importDefault(require("lodash/isEqual"));
var omit_1 = tslib_1.__importDefault(require("lodash/omit"));
var setWith_1 = tslib_1.__importDefault(require("lodash/setWith"));
var ValidatedQuickForm_1 = require("./ValidatedQuickForm");
function Auto(Base) {
    // @ts-expect-error: Mixin class problem.
    var AutoForm = /** @class */ (function (_super) {
        tslib_1.__extends(AutoForm, _super);
        function AutoForm(props) {
            var _this = _super.call(this, props) || this;
            _this.state = tslib_1.__assign(tslib_1.__assign({}, _this.state), { model: props.model });
            return _this;
        }
        AutoForm.prototype.componentDidUpdate = function (prevProps, prevState, snapshot) {
            var model = this.props.model;
            if (!isEqual_1.default(model, prevProps.model)) {
                this.setState({ model: model });
            }
            _super.prototype.componentDidUpdate.call(this, prevProps, prevState, snapshot);
        };
        AutoForm.prototype.getNativeFormProps = function () {
            return omit_1.default(_super.prototype.getNativeFormProps.call(this), ['onChangeModel']);
        };
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        AutoForm.prototype.getModel = function (mode) {
            return this.state.model;
        };
        AutoForm.prototype.onChange = function (key, value) {
            var _this = this;
            _super.prototype.onChange.call(this, key, value);
            this.setState(function (state) { return ({ model: setWith_1.default(clone_1.default(state.model), key, value, clone_1.default) }); }, function () {
                if (_this.props.onChangeModel) {
                    _this.props.onChangeModel(_this.state.model);
                }
            });
        };
        AutoForm.prototype.__reset = function (state) {
            return tslib_1.__assign(tslib_1.__assign({}, _super.prototype.__reset.call(this, state)), { model: this.props.model });
        };
        AutoForm.Auto = Auto;
        AutoForm.displayName = "Auto" + Base.displayName;
        return AutoForm;
    }(Base));
    return AutoForm;
}
exports.Auto = Auto;
exports.AutoForm = Auto(ValidatedQuickForm_1.ValidatedQuickForm);
